### PR TITLE
Fix debounced text input

### LINKED ISSUE #1069: Text input debounce sometimes drops last character of input (Programs)
This seems to be as a result of the recent "useDebouncedInput" hook created for the JSON Forms components.

When typing into a text input quickly, often the last character is lost after debouncing.

### PR DESCRIPTION
**Pull request by clemens-msupply**
Fixes #1069 

Fun issue :) The problem was in `FnUtils.debounce()` (@mark-prins we might have to fix existing usages but a bit more work because the useDebounce hook I am using now works slightly differently...)

The interesting code in `debounce` is:
```typescript
   (...args: Parameters<T>) => {
      if (timer) {
        clearTimeout(timer);
      }
      return new Promise<ReturnType<T>>(resolve => {
        timer = setTimeout(() => {
          const returnValue = callback(...args) as ReturnType<T>;
          resolve(returnValue);
        }, wait);
      });
    };
  }
```

There are two problems here:
1) When the timer gets cleared and the Promise that created this timer isn't resolved yet, the Promise will never resolve. Thus the promise will leak and will float around forever. However, I don't know how much you need to type to create a serious memory leak. We should enable https://typescript-eslint.io/rules/no-floating-promises/ which would have at least indirectly indicated that something might be wrong.
2)  The cause of the bug is that if you call this arrow function too often and too quickly you can run into the race condition where you create multiple promises which all have an active timer. For example, a previous promise hasn't been scheduled yet and thus hasn't called `setTimeout` while a second promise is being spawned (in other works: the second call didn't clear the timer from the previous promise because the previous promise hasn't run yet). These two promises are racing with each other and causing the bug. 

Was able to reproduce the bug easily before but now it seem to be fixed.

Tagging @josh-griffin because I think you wrote it (?) and might find it interesting as well :)

### CODE REVIEW

**Comment by josh-griffin** [comment]
😁 

for the first problem: I think you might be right though if there are no references to the promise I suspect the gc might clean it up, not 100% sure though. The uncertainty here is certainly enough to get rid of the function! 😁 

I don't really remember why I wrote it this way.. I guess as an async-debounce 🤔 ..

Second problem: Nah, the promises callback is executed immediately, so the synchronous code path is -> clear timer, create promise, set new timeout. So, when using a stable debounced function it's not possible to have more than one timer

i.e. this code logs `1,2` not `2,1`:

```
new Promise(() => console.log(1))
console.log(2)
```

The use of the `useDebounceCallback` doesn't keep a stable reference to the debounced function, creating a new function each render, so the promises are *not* clearing each others timers.. just a chain of delayed callbacks which is probably causing some weird behaviour. i.e.

Would also just use a hook library i.e. react-use, rooks or ahooks. Also recommend `useRef` for tracking the last edit time

**Comment by clemens-msupply** [comment]
Yes you are right the promise is executed immediately so 2. is not in issue! Thanks for the feedback yes `useRef` is the better fit here.

Not 100% either but I think the GC would only gc() the promise once `resolve` or `reject` is called. Otherwise the runtime would have to figure out that the promise hangs on a timer and doesn't finish...

**Review by CarlosNZ — CHANGES_REQUESTED** [review]
One weird side effect here.

On this branch, when I create a new patient, the "Create" ("Save") button stays disabled after entering data, so I can't actually save the new patient.

Inspecting the data, I can see it's getting the following validation error, which it's not getting on the original 'feature' branch:

```
validationError must have required property 'isDeceased', must have required property 'socioEconomics'
```

Not sure what's going on there.

**Comment by clemens-msupply** [comment]
Thanks for finding this. There was another "error" bug that was hidden before (see commit) but I have no idea what change causes the creation bug you are seeing... (I thought Richard mentioned that patient creation is working) I fixed the schema and will do another PR.

File: client/packages/programs/src/JsonForms/common/hooks/useDebouncedTextInput.ts [hunk 1]
```diff
@@ -49,8 +48,8 @@ export const useDebouncedTextInput = (
   // performance issue
   const debouncedText = useDebouncedValue(text, 500);
   useEffect(() => {
-    handleChange(error ? undefined : debouncedText);
-  }, [error, debouncedText]);
+    handleChange(debouncedText);
+  }, [handleChange, debouncedText]);
```
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/common/hooks/useDebouncedTextInput.ts:52]: I think the `handleChange` in the dependency array here is a mistake. Seems to be getting into an infinite loop, and not actually saving anything.

When I remove `handleChange` here, it seems to work okay. Was there a particular reason why you included it?

Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/common/hooks/useDebouncedTextInput.ts]: thanks I removed it and this fixed it

**Review by CarlosNZ — APPROVED** [review]
Cool, all seems to be working well now.

Thanks for the fixes

**Pull request merged**